### PR TITLE
Implement code generation

### LINKED ISSUE #15: Implement code generation
Right now `sana_derive` uses `Vm` inside the trait implementation. It should generate native rust code instead.

### PR DESCRIPTION
**Pull request by suhr**
Fixes #15.

### CODE REVIEW

File: sana_core/src/ir.rs [hunk 1]
```diff
@@ -296,62 +296,79 @@ pub enum VmResult<T> {
 }
 
 #[derive(Debug, Clone)]
-pub struct Vm<'code, 'input, T> {
+pub struct Cursor<'input> {
     pub input: &'input str,
-    code: &'code [Op<T>],
     iter: std::str::Chars<'input>,
-    cursor: Option<char>,
+    head: Option<char>,
```
Comment (kpp) [inline sana_core/src/ir.rs:302]: Peek?

Comment (suhr) [reply sana_core/src/ir.rs:302]: `peek` is a verb. And `head` is like a tape head.

File: sana_core/src/ir.rs [hunk 2]
```diff
@@ -296,62 +296,79 @@ pub enum VmResult<T> {
...
     pos: usize,
 }
 
-impl<'code, 'input, T: Clone> Vm<'code, 'input, T> {
-    pub fn new(code: &'code [Op<T>], input: &'input str) -> Self {
+impl<'input> Cursor<'input> {
+    pub fn new(input: &'input str) -> Self {
         let mut iter = input.chars();
         let cursor = iter.next();
```
Comment (kpp) [inline sana_core/src/ir.rs:309]:
```suggestion
        let head = iter.next();
```

File: sana_derive/src/generator.rs [hunk 3]
```diff
@@ -96,42 +97,238 @@ enum Stmt {
...
+        },
+        Stmt::Call(func) => {
+            let func = format_ident!("_l{}", func);
+
+            quote! { return #func(); }
+        },
+        Stmt::Halt =>
+            quote! { return; },
+    }
+}
```
Comment (kpp) [inline sana_derive/src/generator.rs:334]: newline

File: sana_derive/src/generator.rs [hunk 4]
```diff
@@ -96,42 +97,238 @@ enum Stmt {
...
+
+            quote! {
+                let ch = match cursor.head {
+                    Some(ch) => ch,
+                    None => return,
+                };
+
+                match {
+                    #(#bchs),*,
+                    _ => (),
```
Comment (kpp) [inline sana_derive/src/generator.rs:304]:
```suggestion
                    _ => unreachable!(),
```

Comment (suhr) [reply sana_derive/src/generator.rs:304]: `unreachable!()` will be added at the end of the function, not here.

File: sana_derive/src/generator.rs (hunk 3 above)
Comment (suhr) [reply sana_derive/src/generator.rs:334]: Yeah, but it's still WIP.

File: sana_derive/src/generator.rs (hunk 4 above)
Comment (kpp) [reply sana_derive/src/generator.rs:304]: Right.

**Comment by suhr** [comment]
Closed in favor of #43.

**Pull request closed without merging**